converter/converters/pod/affinity.go

```diff
@@ -0,0 +1,398 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if n == nil && p == nil && ap == nil {
+		return nil, nil
+	}
+
+	return &v1.Affinity{
+		NodeAffinity:    n,
+		PodAffinity:     p,
+		PodAntiAffinity: ap,
+	}, nil
+}
+
+type PodAffinity struct {
+	Affinity   string
+	Topology   string
+	Namespaces []string
+}
+
+func splitAffinities(as []types.Affinity) (node []string, pod, antiPod []PodAffinity, err error) {
+	node = []string{}
+	pod = []PodAffinity{}
+	antiPod = []PodAffinity{}
+
+	for _, a := range as {
+		switch {
+		case len(a.NodeAffinity) > 0:
+			node = append(node, a.NodeAffinity)
+		case len(a.PodAffinity) > 0:
+			pod = append(pod, PodAffinity{
+				Affinity:   a.PodAffinity,
+				Topology:   a.Topology,
+				Namespaces: a.Namespaces,
+			})
+		case len(a.PodAntiAffinity) > 0:
+			antiPod = append(antiPod, PodAffinity{
+				Affinity:   a.PodAntiAffinity,
+				Topology:   a.Topology,
+				Namespaces: a.Namespaces,
+			})
+		default:
+			err = fmt.Errorf("unrecognized affinity %#v", a)
+		}
+	}
+
+	return
+}
+
+func podAntiAffinity(as []PodAffinity) (*v1.PodAntiAffinity, error) {
+	if len(as) == 0 {
+		return nil, nil
+	}
+
+	hard, soft, err := podAffinities(as)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hard) == 0 {
+		hard = nil
+	}
+
+	if len(soft) == 0 {
+		soft = nil
+	}
+
+	return &v1.PodAntiAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution:  hard,
+		PreferredDuringSchedulingIgnoredDuringExecution: soft,
+	}, nil
+}
+
+func podAffinity(as []PodAffinity) (*v1.PodAffinity, error) {
+	if len(as) == 0 {
+		return nil, nil
+	}
+
+	hard, soft, err := podAffinities(as)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hard) == 0 {
+		hard = nil
+	}
+
+	if len(soft) == 0 {
+		soft = nil
+	}
+
+	return &v1.PodAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution:  hard,
+		PreferredDuringSchedulingIgnoredDuringExecution: soft,
+	}, nil
+}
+
+func podAffinities(as []PodAffinity) (hard []v1.PodAffinityTerm, soft []v1.WeightedPodAffinityTerm, err error) {
+	hard = []v1.PodAffinityTerm{}
+	soft = []v1.WeightedPodAffinityTerm{}
+	for _, a := range as {
+		segs := strings.Split(a.Affinity, ":")
+		l := len(segs)
+
+		var term *v1.PodAffinityTerm
+		if l < 1 {
+			err = fmt.Errorf("unrecognized PodAffinity %s", a)
+			return
+		} else {
+			term, err = podExprs(segs[0])
+			if err != nil {
+				return
+			}
+
+			term.TopologyKey = a.Topology
+			term.Namespaces = a.Namespaces
+		}
+
+		if l < 2 {
+			hard = append(hard, *term)
+			continue
+		} else {
+			if segs[1] != "soft" {
+				err = fmt.Errorf("unrecognized NodeAffinity term %s", a)
+				return
+			}
+		}
+
+		var weight int64
+		if l < 3 {
+			weight = 1
+		} else {
+			weight, err = strconv.ParseInt(segs[2], 10, 32)
+			if err != nil {
+				return
+			}
+		}
+
+		soft = append(soft, v1.WeightedPodAffinityTerm{
+			Weight:          int32(weight),
+			PodAffinityTerm: *term,
+		})
+	}
+
+	return
+}
+
+func podExprs(s string) (*v1.PodAffinityTerm, error) {
```

Comment:
revertPodExpressions?

Reply:
i decided to call this one parsePodExprs because it's reading the k8s native type directly from a string format.